### PR TITLE
[llama] Add resegment post processing of tokenizer

### PR DESCRIPTION
**Pull request by howard0su**
Try to adress #2023

### CODE REVIEW

**Comment by howard0su** [comment]
@slaren can you give me some test cases?

Also, I rerun the test to see how many tokens would not tokenize to themselves. I got the different result than @vjeux.

**Comment by slaren** [comment]
I would suggest comparing the output with what sentepiecepiece produces. I don't know anything beyond what @vjeux commented.

**Comment by howard0su** [comment]
I read through the code. One thing which concerns me is this line in convert.py:
text = tokenizer.id_to_piece(i).replace("\u2581", " ").encode("utf-8")

when we replace "\u2581" with space, we lose some information.

give the original example in the related issue:
in SPM: 
>>> encoder.encode("▁–")
[29871, 785]

in llama.cpp, token 785 is " ▁–" (there is a leading space). So llama cannot recognize this string is token 785. It will divide this into three chars as many other unicode chars are handling.

Need some insights from @ggerganov why we decide to do the replacement?

**Comment by mqy** [comment]
> Need some insights from @ggerganov why we decide to do the replacement?

@howard0su FYI [FIx parsing single-byte UTF-8 tokens by manually parsing the protobuf](https://github.com/ggerganov/llama.cpp/pull/73) 
Search u2581

TL;DR

**Comment by howard0su** [comment]
Sorry, I cannot find a test case that can hit the new segment code I added.

**Comment by ggerganov** [comment]
I have only brief understanding of how the tokenizer works, so won't be able to help much.
If you observe that your changes match the results from the original implementation then we can merge.

Maybe we can add even more test cases to the `test-tokenizer-0` - why not even generate a bunch of random strings / bytes and verify that we produce the same tokens as the reference implementation.

**Comment by alitariq4589** [comment]
Can one of the admins verify this patch?
